backstory/src/main/java/com/google/sps/images/data/AnnotatedImage.java

```diff
@@ -0,0 +1,119 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.images.data;
+
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import java.lang.IllegalArgumentException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public final class AnnotatedImage {
+  private final byte[] rawImageData;
+  private final List<EntityAnnotation> labelAnnotations;
+
+  /**
+   * AnnotatedImage is an immutable type which represents an image annotated with labels and other
+   * related analytics.
+   *
+   * The image is represented as the byte data in the rawImageData byte array, and the annotated
+   * analytics are represented as labelAnnotations, a list of Vision EntityAnnotation objects, which
+   * each represent individual labels.
+   *
+   * Abstraction Function(rawImageData, labelAnnotations) =
+   * An image represented by the bytes in rawImageData, with a list of labels annotated
+   * to the image, represented by the EntityAnnotation(s) in labelAnnotations.
+   *
+   * Representation Invariant:
+   * rawImageData will always represent the same image, it is immutable. It must be non-null. It
+   * must be non-empty. Once set, labelAnnotations will never change, it is immutable. It must be
+   * non-null.
+   *
+   * Safety from Representation Exposure:
+   * Representation fields are only returned as immutable representations, or (in the case of
+   * rawImageData) as a defensive copy.
+   */
+
+  /**
+   * Asserts that the representation invariants are met
+   * The representation invariants are the conditions which must always
+   * be true about the class representation fields.
+   */
+  private void checkRepresentationInvariantMet() {
+    assert rawImageData != null : "raw image data field is null";
+    assert labelAnnotations != null : "label annotations data field is null";
+    assert rawImageData.length != 0 : "raw image data field is empty";
+  }
+
+  /**
+   * Instantiates the AnnotatedImage object parameterized with a byte array of raw image data and
+   * its labels.
+   *
+   * @param rawImageData The raw image data for the image being represented in this VisionManager
+   *     object. Must be non-empty and non-null.
+   * @param labelAnnotations the preset labels to annotate the image with. Must be non-null.
+   */
+  public AnnotatedImage(byte[] rawImageData, List<EntityAnnotation> labelAnnotations)
+      throws IllegalArgumentException {
+    if (rawImageData == null || labelAnnotations == null || rawImageData.length == 0) {
+      throw new IllegalArgumentException(
+          "Raw image data must be non-null and non-empty, label annotation data must be non-null");
+    }
+
+    this.rawImageData = rawImageData;
+    this.labelAnnotations = labelAnnotations;
+
+    checkRepresentationInvariantMet();
```

Comment:
Is this common practice?

Based on my experiences, I'd favor unit testing this instead, to avoid the need to do this processing in user-facing code.

Reply:
This is a habit I have from a class about writing safe software. I do not know if it is actually common practice - I could have the method there for testing purposes but not actually call it.

Reply:
In theory the rep invariants should never broken.